qiita_db/meta_util.py

```diff
@@ -48,78 +47,117 @@ def _get_data_fpids(constructor, object_id):
     """
-    with qdb.sql_connection.TRN:
+    TRN = qdb.sql_connection.TRN
+    with TRN:
         if user.level == "admin":
+            return True
             # admins have access all files
-            qdb.sql_connection.TRN.add(
-                "SELECT filepath_id FROM qiita.filepath")
-            return set(qdb.sql_connection.TRN.execute_fetchflatten())
-
-        # First, the studies
-        # There are private and shared studies
-        studies = user.user_studies | user.shared_studies
-
-        filepath_ids = set()
-        for study in studies:
-            # Add the sample template files
-            if study.sample_template:
-                filepath_ids.update(
-                    {fid for fid, _ in study.sample_template.get_filepaths()})
-
-            # Add the prep template filepaths
-            for pt in study.prep_templates():
-                filepath_ids.update({fid for fid, _ in pt.get_filepaths()})
-
-            # Add the artifact filepaths
-            for artifact in study.artifacts():
-                filepath_ids.update({fid for fid, _, _ in artifact.filepaths})
-
-        # Next, the public artifacts
-        for artifact in qdb.artifact.Artifact.iter_public():
-            # Add the filepaths of the artifact
-            filepath_ids.update({fid for fid, _, _ in artifact.filepaths})
-
-            # Then add the filepaths of the prep templates
-            for pt in artifact.prep_templates:
-                filepath_ids.update({fid for fid, _ in pt.get_filepaths()})
-
-            # Then add the filepaths of the sample template
-            filepath_ids.update(
-                {fid
-                 for fid, _ in artifact.study.sample_template.get_filepaths()})
-
-        # Next, analyses
-        # Same as before, there are public, private, and shared
-        analyses = qdb.analysis.Analysis.get_by_status('public') | \
-            user.private_analyses | user.shared_analyses
-
-        for analysis in analyses:
-            filepath_ids.update(analysis.all_associated_filepath_ids)
-
-        return filepath_ids
+
+        access = False
+        sql = """SELECT
+            (SELECT count(*) FROM qiita.artifact_filepath
+             WHERE filepath_id = {0}) AS artifact,
+            (SELECT count(*) FROM qiita.sample_template_filepath
+             WHERE filepath_id = {0}) AS sample_info,
+            (SELECT count(*) FROM qiita.prep_template_filepath
+             WHERE filepath_id = {0}) AS prep_info,
+            (SELECT count(*) FROM qiita.job_results_filepath
+             WHERE filepath_id = {0}) AS job_results,
+            (SELECT count(*) FROM qiita.analysis_filepath
+             WHERE filepath_id = {0}) AS analysis""".format(filepath_id)
+        TRN.add(sql)
+
+        arid, sid, pid, jid, anid = TRN.execute_fetchflatten()
+
+        # artifacts
+        if arid:
+            # check the public artifacts
+            public_artifacts = qdb.artifact.Artifact.iter_public()
```

Comment:
I think I have not explained myself correctly on the previous comments. As it is currently written here, you're still iterating over all the artifacts. However, at this point you can now which artifacts this have been seen:
```python
sql = """SELECT artifact_id FROM qiita.artifact_filepath WHERE filepath_id = %s"""
```

Similarly you can do for the other entries. You may be able to modify the above call to return the values in an array so you already have them.